File: java/coraza-java/src/main/java/org/corazawaf/coraza/Transaction.java
package org.corazawaf.coraza;

import org.corazawaf.coraza.internal.swig.coraza;
import org.corazawaf.coraza.internal.swig.coraza_intervention_t;

import java.util.Optional;

/**
 * Represents a single HTTP transaction evaluated by the WAF.
 *
 * <p>Obtain instances via {@link Waf#newTransaction()} or
 * {@link Waf#newTransactionWithId(String)}.
 *
 * <p>The transaction lifecycle follows the Coraza processing phases in order:
 * <ol>
 *   <li>{@link #processConnection}</li>
 *   <li>{@link #processUri}</li>
 *   <li>{@link #addRequestHeader} / {@link #addGetArgs} (optional, before headers)</li>
 *   <li>{@link #processRequestHeaders}</li>
 *   <li>{@link #appendRequestBody} / {@link #requestBodyFromFile} (optional)</li>
 *   <li>{@link #processRequestBody}</li>
 *   <li>{@link #processResponseHeaders}</li>
 *   <li>{@link #addResponseHeader} (optional)</li>
 *   <li>{@link #appendResponseBody} (optional)</li>
 *   <li>{@link #processResponseBody}</li>
 *   <li>{@link #processLogging}</li>
 *   <li>{@link #intervention()} — inspect <em>after</em> the relevant phase</li>
 * </ol>
 *
 * <p>All {@code process*()} and {@code add*()} methods return {@code 0} on success.
 *
 * <p><strong>Thread safety:</strong> a single {@code Transaction} must not be
 * shared across threads.
 *
 * <p><strong>Usage example:</strong>
 * <pre>{@code
 * try (Transaction tx = waf.newTransaction()) {
 *     tx.processConnection("1.2.3.4", 12345, "app.example.com", 443);
 *     tx.processUri("/api/v1/users", "POST", "HTTP/1.1");
 *     tx.addRequestHeader("Content-Type", "application/json");
 *     tx.processRequestHeaders();
 *     tx.appendRequestBody(bodyBytes);
 *     tx.processRequestBody();
 *     tx.processLogging();
 *
 *     Optional<Intervention> it = tx.intervention();
 *     if (it.isPresent()) {
 *         try (Intervention intervention = it.get()) {
 *             return Response.status(intervention.getStatus()).build();
 *         }
 *     }
 * }
 * }</pre>
 */
public final class Transaction implements AutoCloseable {

    private final long tx;
    private boolean freed;

    Transaction(long tx) {
        this.tx = tx;
    }

    // ------------------------------------------------------------------
    // Connection / URI
    // ------------------------------------------------------------------

    /**
     * Records client/server connection metadata.
     *
     * <p>Must be called before {@link #processUri}.
     *
     * @param clientIp   client IP address (IPv4 or IPv6 string)
     * @param clientPort client TCP port
     * @param serverIp   server IP address or hostname
     * @param serverPort server TCP port
     * @return 0 on success
     */
    public int processConnection(
            String clientIp, int clientPort, String serverIp, int serverPort) {
        return coraza.coraza_process_connection(tx, clientIp, clientPort, serverIp, serverPort);
    }

    /**
     * Records the request URI, HTTP method, and protocol version.
     *
     * @param uri         full request URI (may include query string)
     * @param method      HTTP method (e.g. {@code "GET"}, {@code "POST"})
     * @param httpVersion protocol version string (e.g. {@code "HTTP/1.1"})
     * @return 0 on success
     */
    public int processUri(String uri, String method, String httpVersion) {
        return coraza.coraza_process_uri(tx, uri, method, httpVersion);
    }

    // ------------------------------------------------------------------
    // Request headers / query args
    // ------------------------------------------------------------------

    /**
     * Adds a single request header.
     *
     * <p>Call before {@link #processRequestHeaders}.
     *
     * @param name  header name
     * @param value header value
     * @return 0 on success
     */
    public int addRequestHeader(String name, String value) {
        return coraza.coraza_add_request_header(tx, name, name.length(), value, value.length());
    }

    /**
     * Adds a single query-string argument.
     *
     * <p>Can be called before or after {@link #processUri}.
     *
     * @param name  argument name
     * @param value argument value
     * @return 0 on success
     */
    public int addGetArgs(String name, String value) {
        return coraza.coraza_add_get_args(tx, name, value);
    }

    /**
     * Signals that all request headers have been added and runs phase-1 rules.
     *
     * @return 0 on success
     */
    public int processRequestHeaders() {
        return coraza.coraza_process_request_headers(tx);
    }

    // ------------------------------------------------------------------
    // Request body
    // ------------------------------------------------------------------

    /**
     * Appends a chunk of the request body.
     *
     * <p>Call {@link #processRequestBody()} after the last chunk.
     *
     * @param data raw body bytes
     * @return 0 on success
     */
    public int appendRequestBody(byte[] data) {
        return coraza.coraza_append_request_body(tx, data);
    }

    /**
     * Signals end-of-request-body and runs phase-2 rules.
     *
     * @return 0 on success
     */
    public int processRequestBody() {
        return coraza.coraza_process_request_body(tx);
    }

    /**
     * Loads the request body from a file on disk.
     *
     * <p>Useful for large uploads already spooled to a temporary file.
     *
     * @param path absolute path to the body file
     * @return 0 on success
     */
    public int requestBodyFromFile(String path) {
        return coraza.coraza_request_body_from_file(tx, path);
    }

    // ------------------------------------------------------------------
    // Response
    // ------------------------------------------------------------------

    /**
     * Records the upstream response status and runs phase-3 rules.
     *
     * @param status      HTTP status code from the upstream response
     * @param httpVersion protocol version string (e.g. {@code "HTTP/1.1"})
     * @return 0 on success
     */
    public int processResponseHeaders(int status, String httpVersion) {
        return coraza.coraza_process_response_headers(tx, status, httpVersion);
    }

    /**
     * Adds a single response header.
     *
     * @param name  header name
     * @param value header value
     * @return 0 on success
     */
    public int addResponseHeader(String name, String value) {
        return coraza.coraza_add_response_header(tx, name, name.length(), value, value.length());
    }

    /**
     * Appends a chunk of the response body.
     *
     * @param data raw body bytes
     * @return 0 on success
     */
    public int appendResponseBody(byte[] data) {
        return coraza.coraza_append_response_body(tx, data);
    }

    /**
     * Signals end-of-response-body and runs phase-4 rules.
     *
     * @return 0 on success
     */
    public int processResponseBody() {
        return coraza.coraza_process_response_body(tx);
    }

    /**
     * Updates the HTTP status code after the response has been processed.
     *
     * @param status new HTTP status code
     */
    public void updateStatusCode(int status) {
        coraza.coraza_update_status_code(tx, status);
    }

    // ------------------------------------------------------------------
    // Finalization
    // ------------------------------------------------------------------

    /**
     * Runs phase-5 (logging) rules and flushes the audit log.
     *
     * <p>Always call this before {@link #intervention()} and before {@link #free()}.
     *
     * @return 0 on success
     */
    public int processLogging() {
        return coraza.coraza_process_logging(tx);
    }

    /**
     * Returns the current WAF intervention wrapped in an {@link Optional}.
     *
     * <p>Returns {@link Optional#empty()} when no rule matched.  When present,
     * the {@link Intervention} <strong>must</strong> be closed after use to
     * release native memory:
     *
     * <pre>{@code
     * tx.intervention().ifPresent(it -> {
     *     try (it) {
     *         response.setStatus(it.getStatus());
     *     }
     * });
     * }</pre>
     *
     * @return an {@code Optional} containing the intervention, or empty
     */
    public Optional<Intervention> intervention() {
        coraza_intervention_t raw = coraza.coraza_intervention(tx);
        return Optional.ofNullable(raw).map(Intervention::new);
    }

    /**
     * Releases the transaction.  Idempotent — safe to call multiple times.
     *
     * @return 0 on success
     */
    public int free() {
        if (freed) return 0;
        freed = true;
        return coraza.coraza_free_transaction(tx);
    }

    /** Calls {@link #free()}. */
    @Override
    public void close() {
        free();
    }
}

File: nodejs/src/transaction.ts
// Copyright 2024 OWASP Coraza contributors
// SPDX-License-Identifier: Apache-2.0

import type { CorazaExports } from "./loader.js";
import { Memory } from "./memory.js";
import { type Intervention, readIntervention } from "./intervention.js";

/** A single HTTP transaction evaluated by the WAF. Not thread-safe. */
export class Transaction {
  readonly #exports: CorazaExports;
  readonly #memory: Memory;
  readonly #handle: number;
  #freed = false;

  /** @internal */
  constructor(exports: CorazaExports, memory: Memory, handle: number) {
    this.#exports = exports;
    this.#memory = memory;
    this.#handle = handle;
  }

  #check(): void {
    if (this.#freed) throw new Error("Transaction already freed");
  }

  #intervention(rc: number): Intervention | null {
    if (rc < 0)
      throw new Error(`coraza WASM error (code ${rc})`);
    if (rc === 0) return null;
    const itHandle = this.#exports.coraza_intervention(this.#handle);
    if (itHandle === 0) return null;
    return readIntervention(this.#exports, this.#memory, itHandle);
  }

  /** Record connection metadata. */
  processConnection(
    clientIp: string,
    clientPort: number,
    serverIp: string,
    serverPort: number,
  ): void {
    this.#check();
    this.#memory.withStrings(clientIp, serverIp, (cPtr, cLen, sPtr, sLen) =>
      this.#exports.coraza_process_connection(
        this.#handle,
        cPtr,
        cLen,
        clientPort,
        sPtr,
        sLen,
        serverPort,
      ),
    );
  }

  /** Record the request URI, method, and HTTP version. */
  processUri(uri: string, method: string, httpVersion: string): void {
    this.#check();
    const u = this.#memory.writeString(uri);
    const m = this.#memory.writeString(method);
    const p = this.#memory.writeString(httpVersion);
    try {
      this.#exports.coraza_process_uri(
        this.#handle,
        u.ptr,
        u.len,
        m.ptr,
        m.len,
        p.ptr,
        p.len,
      );
    } finally {
      this.#memory.free(u.ptr);
      this.#memory.free(m.ptr);
      this.#memory.free(p.ptr);
    }
  }

  /** Add a request header. */
  addRequestHeader(name: string, value: string): void {
    this.#check();
    this.#memory.withStrings(name, value, (nPtr, nLen, vPtr, vLen) =>
      this.#exports.coraza_add_request_header(
        this.#handle,
        nPtr,
        nLen,
        vPtr,
        vLen,
      ),
    );
  }

  /** Add a GET query-string argument. */
  addGetArgument(name: string, value: string): void {
    this.#check();
    this.#memory.withStrings(name, value, (nPtr, nLen, vPtr, vLen) =>
      this.#exports.coraza_add_get_args(
        this.#handle,
        nPtr,
        nLen,
        vPtr,
        vLen,
      ),
    );
  }

  /** Evaluate request headers. Returns an intervention if a rule matched. */
  processRequestHeaders(): Intervention | null {
    this.#check();
    const rc = this.#exports.coraza_process_request_headers(this.#handle);
    return this.#intervention(rc);
  }

  /**
   * Write a chunk of the request body. Returns an intervention if a body limit
   * was exceeded and action is to reject.
   */
  appendRequestBody(data: Uint8Array | string): Intervention | null {
    this.#check();
    const bytes =
      typeof data === "string" ? new TextEncoder().encode(data) : data;
    const rc = this.#memory.withBytes(bytes, (ptr, len) =>
      this.#exports.coraza_append_request_body(this.#handle, ptr, len),
    );
    return this.#intervention(rc);
  }

  /** Evaluate the full request body. */
  processRequestBody(): Intervention | null {
    this.#check();
    const rc = this.#exports.coraza_process_request_body(this.#handle);
    return this.#intervention(rc);
  }

  /** Stream a file into the request body buffer and evaluate it. */
  requestBodyFromFile(path: string): Intervention | null {
    this.#check();
    const rc = this.#memory.withString(path, (ptr, len) =>
      this.#exports.coraza_request_body_from_file(this.#handle, ptr, len),
    );
    return this.#intervention(rc);
  }

  /** Add a response header. */
  addResponseHeader(name: string, value: string): void {
    this.#check();
    this.#memory.withStrings(name, value, (nPtr, nLen, vPtr, vLen) =>
      this.#exports.coraza_add_response_header(
        this.#handle,
        nPtr,
        nLen,
        vPtr,
        vLen,
      ),
    );
  }

  /** Evaluate response headers. */
  processResponseHeaders(status: number, proto: string): Intervention | null {
    this.#check();
    const rc = this.#memory.withString(proto, (ptr, len) =>
      this.#exports.coraza_process_response_headers(
        this.#handle,
        status,
        ptr,
        len,
      ),
    );
    return this.#intervention(rc);
  }

  /** Write a chunk of the response body. */
  appendResponseBody(data: Uint8Array | string): Intervention | null {
    this.#check();
    const bytes =
      typeof data === "string" ? new TextEncoder().encode(data) : data;
    const rc = this.#memory.withBytes(bytes, (ptr, len) =>
      this.#exports.coraza_append_response_body(this.#handle, ptr, len),
    );
    return this.#intervention(rc);
  }

  /** Evaluate the full response body. */
  processResponseBody(): Intervention | null {
    this.#check();
    const rc = this.#exports.coraza_process_response_body(this.#handle);
    return this.#intervention(rc);
  }

  /**
   * Update the transaction's tracked response status code.
   * Used by reverse-proxy integrations when the upstream status is known.
   */
  updateStatusCode(code: number): void {
    this.#check();
    this.#exports.coraza_update_status_code(this.#handle, code);
  }

  /** Run the logging phase. Should be called once per transaction lifecycle. */
  processLogging(): void {
    this.#check();
    this.#exports.coraza_process_logging(this.#handle);
  }

  /**
   * Return the current intervention without advancing the transaction state.
   * Null if the transaction has not been interrupted.
   */
  intervention(): Intervention | null {
    this.#check();
    const itHandle = this.#exports.coraza_intervention(this.#handle);
    if (itHandle === 0) return null;
    return readIntervention(this.#exports, this.#memory, itHandle);
  }

  /**
   * Release the native transaction handle.
   * Calls processLogging() automatically if not already called.
   * Must be called exactly once; subsequent calls are no-ops.
   */
  free(): void {
    if (!this.#freed) {
      this.#freed = true;
      this.#exports.coraza_free_transaction(this.#handle);
    }
  }

  [Symbol.dispose](): void {
    this.free();
  }
}
